server/src/main/java/org/apache/gravitino/server/web/rest/MetadataObjectTagOperations.java

```diff
@@ -187,18 +189,34 @@ public Response listTagsForMetadataObject(
               parentObject = MetadataObjects.parent(parentObject);
             }
 
+            // Deduplicate by tag name, prefer direct(inherited==false) over inherited(true)
+            Map<String, TagDTO> byName = new LinkedHashMap<>();
+            for (TagDTO t : tags) {
+              TagDTO prev = byName.get(t.name());
+              if (prev == null) {
+                byName.put(t.name(), t);
+              } else {
+                boolean prevDirect = !prev.inherited().orElse(false);
+                boolean currDirect = !t.inherited().orElse(false);
+                if (!prevDirect && currDirect) {
+                  byName.put(t.name(), t);
+                }
+              }
+            }
+
```

Comment:
There may be a simplified way to do this. e.g. Add all direct tags, then add any inherited tags that have not all been added.